internal/graph/check.go

```diff
@@ -0,0 +1,536 @@
+
+	typesys, ok := typesystem.TypesystemFromContext(parentctx)
+	if !ok {
+		panic("typesystem missing in context")
+	}
+
+	return func(ctx context.Context) (*openfgapb.CheckResponse, error) {
+
+		storeID := req.GetStoreID()
+		tk := req.GetTupleKey()
+
+		fn1 := func(ctx context.Context) (*openfgapb.CheckResponse, error) {
+			t, err := c.ds.ReadUserTuple(ctx, storeID, tk)
+			if err != nil {
+				if errors.Is(err, storage.ErrNotFound) {
+					return &openfgapb.CheckResponse{Allowed: false}, nil
+				}
+
+				return &openfgapb.CheckResponse{Allowed: false}, err
+			}
+
+			// filter out invalid tuples yielded by the database query
+			err = validation.ValidateTuple(typesys, tk)
+
+			if t != nil && err == nil {
+				return &openfgapb.CheckResponse{Allowed: true}, nil
+			}
+
+			return &openfgapb.CheckResponse{Allowed: false}, nil
+		}
+
+		fn2 := func(ctx context.Context) (*openfgapb.CheckResponse, error) {
+			iter, err := c.ds.ReadUsersetTuples(ctx, storeID, tk)
+			if err != nil {
+				return &openfgapb.CheckResponse{Allowed: false}, err
+			}
+			defer iter.Stop()
+
+			// filter out invalid tuples yielded by the database iterator
+			filteredIter := storage.NewFilteredTupleKeyIterator(
+				storage.NewTupleKeyIteratorFromTupleIterator(iter),
+				validation.FilterInvalidTuples(typesys),
+			)
+			defer filteredIter.Stop()
+
+			var handlers []CheckHandlerFunc
+			for {
+				t, err := filteredIter.Next(ctx)
+				if err != nil {
+					if errors.Is(err, storage.ErrIteratorDone) {
+						break
+					}
+
+					return &openfgapb.CheckResponse{Allowed: false}, err
+				}
+
+				usersetObject, usersetRelation := tuple.SplitObjectRelation(t.GetUser())
+
+				// for 1.0 models, if the user is '*' then we're done searching
+				if usersetObject == tuple.Wildcard && typesys.GetSchemaVersion() == typesystem.SchemaVersion1_0 {
+					return &openfgapb.CheckResponse{Allowed: true}, nil
+				}
+
+				// for 1.1 models, if the user value is a typed wildcard and the type of the wildcard
+				// matches the target user objectType, then we're done searching
+				if tuple.IsTypedWildcard(usersetObject) && typesys.GetSchemaVersion() == typesystem.SchemaVersion1_1 {
+
+					wildcardType := tuple.GetType(usersetObject)
+
+					if tuple.GetType(tk.GetUser()) == wildcardType {
+						return &openfgapb.CheckResponse{Allowed: true}, nil
+					}
+
+					return &openfgapb.CheckResponse{Allowed: false}, nil
+				}
+
+				if usersetRelation != "" {
+					handlers = append(handlers, c.dispatch(
+						ctx,
+						&dispatcher.DispatchCheckRequest{
+							StoreID:              storeID,
+							AuthorizationModelID: req.GetAuthorizationModelID(),
+							TupleKey:             tuple.NewTupleKey(usersetObject, usersetRelation, tk.GetUser()),
+							ResolutionMetadata: &dispatcher.ResolutionMetadata{
+								Depth: req.GetResolutionMetadata().Depth - 1,
+							},
+						}))
+				}
+
+			}
+
+			if len(handlers) > 0 {
```

Comment:
Small nitpick, but I think it would read a bit nicer if the `if` handled the "negative" case of `len(handlers) == 0`. Similarly elsewhere. But up to you.

Reply:
👍 done.